flow/src/main/java/com/sequenceiq/flow/core/Flow2Handler.java

```diff
@@ -185,19 +185,42 @@ public void cancelFlow(Long stackId, String flowId) throws TransactionExecutionE
         }
     }
 
-    private void finalizeFlow(FlowParameters flowParameters, String flowChainId, Long stackId, Map<Object, Object> contextParams)
+    private void finalizeFlow(FlowParameters flowParameters, String flowChainId, Long resourceId, Map<Object, Object> contextParams)
             throws TransactionExecutionException {
         String flowId = flowParameters.getFlowId();
         LOGGER.debug("flow finalizing arrived: id: {}", flowId);
-        flowLogService.close(stackId, flowId);
+        flowLogService.close(resourceId, flowId);
         Flow flow = runningFlows.remove(flowId);
+        Optional<Runnable> finalizerCallback = createFinalizerCallback(flow, resourceId);
         flowStatCache.remove(flowId, flowChainId == null && !flow.isFlowFailed());
         if (flowChainId != null) {
             if (flow.isFlowFailed()) {
                 flowChains.removeFullFlowChain(flowChainId, false);
+                finalizerCallback.ifPresent(Runnable::run);
             } else {
-                flowChains.triggerNextFlow(flowChainId, flowParameters.getFlowTriggerUserCrn(), contextParams, flowParameters.getFlowOperationType());
+                flowChains.triggerNextFlow(flowChainId, flowParameters.getFlowTriggerUserCrn(), contextParams, flowParameters.getFlowOperationType(),
+                        finalizerCallback);
             }
+        } else {
+            finalizerCallback.ifPresent(Runnable::run);
+        }
+    }
+
+    private Optional<Runnable> createFinalizerCallback(Flow flow, Long resourceId) {
```

Comment:
If I understand correctly, this magic with Runnable is necessary because of the error handling of the onFinalize method calls?

Reply:
Yes, unfortunately the flow logic is separated in a really messy way into different classes which makes sense only if those are working together. I didn't try to fix methods in `FlowChans`, `Flow2Handler` (to have a more clear API)... so I could only come up with this to pass a method as an argument. One more things is that I didn't want to have a failure because of this `onFinalize` method fails, therefore I put a try-catch around it. I agree that this is not a clear solution but I couldn't figure out a better way without duplicating code.

Reply:
It makes sense. Maybe if FlowFinalizerCallback would be an abstract class which implements the exception handling logic around the actual finalize method, something like this? 

```
    public abstract class FlowFinalizerCallback {

        public abstract void doOnFinalize(Long resourceId);

        public void onFinalize(Long resourceId) {
            try {
                doOnFinalize(resourceId);
            } catch (Exception e) {
                // logging, etc.
            }
        }
    }
```